<center>
  <a href="MLSD-01-GettingStarted.ipynb" target="_self">Getting Started</a> | <a href="./">Content Page</a> | <a href="MLSD-02-DataPreprocessing-B.ipynb">Data Preprocessing B | <a href="MLSD-02-DataPreprocessing-Ex-1.ipynb">Data Preprocessing Exercise</a>
</center>

# <center>DATA PREPROCESSING A</center>

<center><b>Copyright &copy 2023 by DR DANNY POO</b><br> e:dannypoo@nus.edu.sg<br> w:drdannypoo.com</center><br>

# Data Preprocessing 
<b>Dataset</b>: Titanic data set.<br>
<b>Tasks</b>: 
- To read in and explore data set.
- To drop columns that are not useful.
- To drop rows with missing values.
- To encode all categorical labels to column vectors with binary values. 
- To fill missing data with interpolated values. 

## Read in and Explore Data Set

In [1]:
# Import libraries
import numpy as np
import pandas as pd

In [2]:
# Read in data
df = pd.read_csv('./data/titanic/train.csv')
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
# Examine data format
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


**Observations**:
- There are 891 total rows 
- Age shows only 714 (which means we’re missing some data)
- Embarked is missing two rows
- Cabin is missing a lot as well
- Object data types are non-numeric so we have to find a way to encode them to numerical values

## Drop Columns that are not Useful

**To drop**:
- Name
- Ticket
- Cabin

In [4]:
# Drop columns Name, Ticket, Cabin
cols = ['Name', 'Ticket', 'Cabin']
df = df.drop(cols, axis=1)
df.head()

PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0            1         0       3    male  22.0      1      0   7.2500        S
1            2         1       1  female  38.0      1      0  71.2833        C
2            3         1       3  female  26.0      0      0   7.9250        S
3            4         1       1  female  35.0      1      0  53.1000        S
4            5         0       3    male  35.0      0      0   8.0500        S

## Drop Rows with Missing Values
Missing values may arise due to:
- participants' failure to respond to some questions, not knowing the correct response, or being unwilling to answer
- error made during the data entry process.

Missing values may:
- lead to bias in the estimation of the parameters
- compromise the accuracy of the machine learning models.

Resulting in drawing wrong conclusions about data.

**To drop**:
- all rows that have missing values (NaNs)

In [5]:
# Before we drop all rows that have missing values, let's duplicate df
dfDup = df

# Drop all rows that have missing values (NaNs)
df = df.dropna()
df.info()## Creating Dummy Variables

<class 'pandas.core.frame.DataFrame'>
Int64Index: 712 entries, 0 to 890
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  712 non-null    int64  
 1   Survived     712 non-null    int64  
 2   Pclass       712 non-null    int64  
 3   Sex          712 non-null    object 
 4   Age          712 non-null    float64
 5   SibSp        712 non-null    int64  
 6   Parch        712 non-null    int64  
 7   Fare         712 non-null    float64
 8   Embarked     712 non-null    object 
dtypes: float64(2), int64(5), object(2)
memory usage: 55.6+ KB


**Observations**:
- After dropping rows with missing values, data set is reduced to 712 rows from 891.
- We are wasting data.

## Encode Categorical Values

Pclass, Sex and Embarked are categorical.<br>
Need to encode them into integer columns.

**To convert**:
- Pclass, Sex and Embarked to columns

In [6]:
# Restore df to original data set
df = dfDup

# Convert Pclass, Sex and Embarked with dummy variables
dummies = []
cols = ['Pclass', 'Sex', 'Embarked']
for col in cols:
   dummies.append(pd.get_dummies(df[col]))

titanic_dummies = pd.concat(dummies, axis=1)
titanic_dummies

1  2  3  female  male  C  Q  S
0    0  0  1       0     1  0  0  1
1    1  0  0       1     0  1  0  0
2    0  0  1       1     0  0  0  1
3    1  0  0       1     0  0  0  1
4    0  0  1       0     1  0  0  1
..  .. .. ..     ...   ... .. .. ..
886  0  1  0       0     1  0  0  1
887  1  0  0       1     0  0  0  1
888  0  0  1       1     0  0  0  1
889  1  0  0       0     1  1  0  0
890  0  0  1       0     1  0  1  0

[891 rows x 8 columns]

**Observations**:
- Columns 1, 2, 3 represent the passenger class (Pclass)
- Columns female, male represent Sex
- Columns C, Q, S represent Embarked

In [7]:
# Concatenate them to the original data frame, column-wise
df = pd.concat((df,titanic_dummies), axis=1)
df.head()

PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch     Fare  \
0            1         0       3    male  22.0      1      0   7.2500   
1            2         1       1  female  38.0      1      0  71.2833   
2            3         1       3  female  26.0      0      0   7.9250   
3            4         1       1  female  35.0      1      0  53.1000   
4            5         0       3    male  35.0      0      0   8.0500   

  Embarked  1  2  3  female  male  C  Q  S  
0        S  0  0  1       0     1  0  0  1  
1        C  1  0  0       1     0  1  0  0  
2        S  0  0  1       1     0  0  0  1  
3        S  1  0  0       1     0  0  0  1  
4        S  0  0  1       0     1  0  0  1

In [8]:
# Drop the redundant Pclass, Sex and Embarked columns
df = df.drop(['Pclass', 'Sex', 'Embarked'], axis=1)
df.head()

PassengerId  Survived   Age  SibSp  Parch     Fare  1  2  3  female  male  \
0            1         0  22.0      1      0   7.2500  0  0  1       0     1   
1            2         1  38.0      1      0  71.2833  1  0  0       1     0   
2            3         1  26.0      0      0   7.9250  0  0  1       1     0   
3            4         1  35.0      1      0  53.1000  1  0  0       1     0   
4            5         0  35.0      0      0   8.0500  0  0  1       0     1   

   C  Q  S  
0  0  0  1  
1  1  0  0  
2  0  0  1  
3  0  0  1  
4  0  0  1

In [9]:
# The new data frame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Age          714 non-null    float64
 3   SibSp        891 non-null    int64  
 4   Parch        891 non-null    int64  
 5   Fare         891 non-null    float64
 6   1            891 non-null    uint8  
 7   2            891 non-null    uint8  
 8   3            891 non-null    uint8  
 9   female       891 non-null    uint8  
 10  male         891 non-null    uint8  
 11  C            891 non-null    uint8  
 12  Q            891 non-null    uint8  
 13  S            891 non-null    uint8  
dtypes: float64(2), int64(4), uint8(8)
memory usage: 48.9 KB


## Fill Missing Data with Interpolated Values

Age has lots of missing values.<br>
Let's `interpolate()` all the ages and fill those missing Age values.<br>
pandas has an `interpolate()` function that will replace all missing NaNs to interpolated values.

In [10]:
# interpolate Age
df['Age'] = df['Age'].interpolate()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Age          891 non-null    float64
 3   SibSp        891 non-null    int64  
 4   Parch        891 non-null    int64  
 5   Fare         891 non-null    float64
 6   1            891 non-null    uint8  
 7   2            891 non-null    uint8  
 8   3            891 non-null    uint8  
 9   female       891 non-null    uint8  
 10  male         891 non-null    uint8  
 11  C            891 non-null    uint8  
 12  Q            891 non-null    uint8  
 13  S            891 non-null    uint8  
dtypes: float64(2), int64(4), uint8(8)
memory usage: 48.9 KB


**Observations**:
- Columns Age is now interpolated with imputed new values.

### Convert the DataFrame to Numpy

All data are now integers.<br>
Let's prepare the data for machine learning models.<br>
X will be input set with 14 features.<br>
y will be output with Survived column.

In [11]:
# Convert dataframe from pandas to Numpy and assign input and output.
X = df.values
y = df['Survived'].values

In [12]:
# Remove Survived from X
X = np.delete(X, 1, axis=1)

### Divide the Data Set into Training Data and Test Data

Split data set: allocate 70% for training and 30% for test using sciket `model_selection`.

In [13]:
# Import libraries
from sklearn.model_selection import train_test_split

# Split dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

In [14]:
# Print
print("\nX_train\n", X_train)
print("\nX_test\n", X_test)
print("\ny_train\n", y_train)
print("\ny_test\n", y_test)


X_train
 [[858.  51.   0. ...   0.   0.   1.]
 [ 53.  49.   1. ...   1.   0.   0.]
 [387.   1.   5. ...   0.   0.   1.]
 ...
 [630.  53.   0. ...   0.   1.   0.]
 [560.  36.   1. ...   0.   0.   1.]
 [685.  60.   1. ...   0.   0.   1.]]

X_test
 [[496.          37.5          0.         ...   1.           0.
    0.        ]
 [649.          39.5          0.         ...   0.           0.
    1.        ]
 [279.           7.           4.         ...   0.           1.
    0.        ]
 ...
 [621.          27.           1.         ...   1.           0.
    0.        ]
 [787.          18.           0.         ...   0.           0.
    1.        ]
 [ 65.          12.33333333   0.         ...   1.           0.
    0.        ]]

y_train
 [1 1 0 0 0 1 0 0 0 1 1 0 0 1 0 1 0 0 0 0 0 0 1 0 1 1 1 0 0 0 1 0 1 0 0 1 1
 1 0 0 1 0 1 0 0 0 0 1 0 1 0 1 0 1 1 1 0 0 0 0 0 0 1 0 1 0 0 0 1 0 1 0 0 0
 1 0 1 1 1 0 0 0 1 1 0 0 1 0 1 0 0 0 0 1 1 1 0 1 0 1 0 1 0 0 1 0 0 0 0 0 1
 0 1 0 1 0 0 0 0 0 1 0 0 1 0 0 0 1 0 0

<center>
  <a href="MLSD-01-GettingStarted.ipynb" target="_self">Getting Started</a> | <a href="./">Content Page</a> | <a href="MLSD-02-DataPreprocessing-B.ipynb">Data Preprocessing B | <a href="MLSD-02-DataPreprocessing-Ex-1.ipynb">Data Preprocessing Exercise</a>
</center>